### ISSUE TITLE
Orphan dotnet processes from tasks.json Attach to C# Functions (macOS)

### ISSUE THREAD

**Issue by ryanspletzer** [microsoft/vscode-azurefunctions#992]
macOS Mojave 10.14.3

Repro steps:

- Install latest VS Code fresh as well as .NET Core SDK and Azure Functions Core Tools (via homebrew). Optionally, add the C# VS Code extension.
- Add Azure Functions extension
- Create a helloworld project with a single function
- Run Attach to C# Functions task
- Note the present `dotnet` process in addition to `func`
- Stop debugging
- Note the still existing `dotnet` process

If you run through the above again, you'll wind up with another orphaned dotnet process.

Is this intentional / am I possibly doing something wrong?

(I didn't post this for Windows because I can't actually get it attached at all there properly on a fresh VM -- separate issue.)

Thank you for your time!!

**Comment by EricJizbaMSFT** [comment]
Definitely not intentional. If you run `func host start` from the command line (outside of VS Code) - do you get the same behavior? Our extension is stopping the func cli when you stop debugging, but I would think it's up to the func cli to clean up its child processes. If you see the same behavior outside VS Code, I would suggest filing an issue here instead: https://github.com/Azure/azure-functions-core-tools

**Comment by ryanspletzer** [comment]
Hi Eric, I had the same thought. When I do `func host start` or `func start` outside of VS Code, there is no orphaned dotnet process, even while func is running. If you execute func in the context of the project directory where the .csproj is, func does indeed invoke dotnet to re-build first, then starts the runtime. (If you cd into the bin and execute, no re-build is needed.)

I believe it's only happening in the context of VS Code + Azure Functions extension via the tasks.json definition for dotnet build. Somehow the interaction between the "build" task and the next task for "func" where it attaches to the process leaves a dotnet process sitting there... I think it's the dotnet process that was invoked for the build prior to calling func host start in the next task. (To be clear, this happens whether I hit "Stop Debugging" in the Debug menu, *or* if I hit CTRL + C to stop func -- I think that the process is getting orphaned sooner than either of those things.)

I could put up sample code, but even a normal helloworld app with no functions in it (generated via the 
Azure Functions extension) and all the default tasks.json etc. files in the .vscode folder from the Azure Functions extension is enough to repro.

**Comment by ryanspletzer** [comment]
tl;dr from above -- I don't think in this scenario in VS Code + Azure Functions extension that dotnet is a child process of func, I think it's called before that and func doesn't invoke dotnet in this scenario in vs code (because it's invoked a different, prior task -- but please correct me if I'm wrong).

**Comment by EricJizbaMSFT** [comment]
Huh well if the dotnet process is from the build, then we have no control over that. We write down your "tasks.json" when you first create a project, but at that point you can configure it however you want and VS Code itself (not our extension) controls the task system.

What version of `dotnet` do you have? I'm on 2.1.403 and I see these processes _during_ the build (macOS Mojave 10.14.3 as well):
![screen shot 2019-02-15 at 11 18 58 am](https://user-images.githubusercontent.com/11282622/52879320-adb57200-3113-11e9-86b1-21fd13745687.png)

But once `func host start` happens I see nothing:
![screen shot 2019-02-15 at 11 19 08 am](https://user-images.githubusercontent.com/11282622/52879326-b148f900-3113-11e9-8681-5e875adc23c6.png)

(I was using "CSharp: List processes for attach" from the command palette to list those)

**Comment by ryanspletzer** [comment]
Interesting. Re: dotnet version:

![image](https://user-images.githubusercontent.com/1481108/52880965-6047fd80-3131-11e9-82b4-31aec5e40b79.png)

The tasks.json I have is the OOTB one generated by the extension for a project:

```
{
  "version": "2.0.0",
  "tasks": [
    {
      "label": "clean",
      "command": "dotnet clean",
      "type": "shell",
      "problemMatcher": "$msCompile"
    },
    {
      "label": "build",
      "command": "dotnet build",
      "type": "shell",
      "dependsOn": "clean",
      "group": {
        "kind": "build",
        "isDefault": true
      },
      "problemMatcher": "$msCompile"
    },
    {
      "label": "clean release",
      "command": "dotnet clean --configuration Release",
      "type": "shell",
      "problemMatcher": "$msCompile"
    },
    {
      "label": "publish",
      "command": "dotnet publish --configuration Release",
      "type": "shell",
      "dependsOn": "clean release",
      "problemMatcher": "$msCompile"
    },
    {
      "label": "runFunctionsHost",
      "type": "shell",
      "dependsOn": "build",
      "options": {
        "cwd": "${workspaceFolder}/bin/Debug/netcoreapp2.1"
      },
      "command": "func host start",
      "isBackground": true,
      "problemMatcher": "$func-watch"
    }
  ]
}
```

This is what I see *during* build and also when `func host start` happens (i.e. dotnet doesn't go away):

![image](https://user-images.githubusercontent.com/1481108/52882301-3a245c80-3135-11e9-9e93-b08b9fef3b51.png)

I wish I could spin up macOS in a VM to try in a clean machine like I can with Windows. :)

**Comment by EricJizbaMSFT** [comment]
I'm going to close this one as external. From the above discussion it sounds like the orphaned dotnet processes are coming from build, not running the func host. In addition, I can't repro on macOS. @ryanspletzer if this is still bothering you I suggest trying to get a simple repro with just the dotnet build tasks (aka with a simple console application instead of a functions project) and filing an issue either on VS Code (since the problem could be with their task system) or with dotnet itself.
